### PR TITLE
[WIP] Add logging support

### LINKED ISSUE #27958: Add built-in logging classes / capabilities
**Godot version:** 3.1

**Issue description:**
I do use logging a lot to track things in the game and investigate bugs. But I don't believe that I do that in an efficient way (or that there's such way at all for GDScript). I believe this is a very basic feature, but I've failed to find something related to that.

**What I expect from logger**
This is a class that allows and supports:

* customizing logging format. Say, similar to `printf` special characters, for example:
  - `%t` would mean current time like `2012-01-30T10:20:30Z` or whatever
  - `%n` would mean name of the logger
  - `%t` would mean name of thread this logger is triggered from
  - etc
* log levels. The most important are:
  - ERROR
  - WARN
  - INFO
  - DEBUG
  - TRACE
* logging destinations:
  - console
  - file
  - what else?
* remove logging-related pieces of code if the game is not in "debug" mode. It would be fine to have it implemented like directives in C, say:
```GDScript
##ifdef DEBUG
logger.info("some interesting stuff here")
##endif
```
* log files rotation policy (optional, I don't believe that there would be developers with need in this)

Currently, I'm doing this with my own `Logger` class written in GDScript and handling all that levels. But even so, when I need to dig into an issue and set logging level to TRACE (i.e. I see LOTS of logs with every little detail happening in my game) I get irritating message:
```
output overflow, print less text
```
But, hell, I do need to print lots of text!

### PR DESCRIPTION
**Pull request by mitchcurtis**
This patch adds a debug_log() function (log() was taken; the name is up for discussion) to GDScript. The goals are to:

- Allow logging that is only executed in debug builds.
- Allow logging to different categories, which can be enabled/disabled in e.g. project settings.
- Allow specifying a message pattern (e.g. in project settings) to enable things like timestamps to be included in the logged messages (see e.g. https://doc.qt.io/qt-5/qtglobal.html#qSetMessagePattern).
- Allow logging to either the console or a file.

Closes #27958.

### CODE REVIEW

**Comment by mitchcurtis** [comment]
Opening this early so I can get early feedback. What do you think about the goals of this patch?

**Comment by fire** [comment]
You may be able to use and improve on this function for timestamps. get_iso_date_time

**Comment by Calinou** [comment]
As a stretch goal, maybe we could have a setting to output all logging messages as JSON. This would be useful to feed a dedicated server's console output to a log aggregation tool such as [Loki](https://github.com/grafana/loki).

**Comment by mitchcurtis** [comment]
> As a stretch goal, maybe we could have a setting to output all logging messages as JSON. This would be useful to feed a dedicated server's console output to a log aggregation tool such as [Loki](https://github.com/grafana/loki).

Sounds fairly straightforward, assuming there's already a way to write JSON to a text file?

**Comment by Calinou** [comment]
@mitchcurtis You would just need to print the JSON to standard output (one JSON object per line). Writing to a file should be handled by writing that standard output to a file, as is already done by the **Logging > File Logging > Enable File Logging** project setting.

**Comment by mitchcurtis** [comment]
Is there a way I can suppress the notifications for reviewers until I'm ready? I still have a lot of work to do... perhaps I should have just kept this as a branch on my fork. :x

**Comment by mitchcurtis** [comment]
@bojidar-bg do you know why I'd be getting this warning:

> W 0:00:00:0447   Standalone expression (the line has no effect).
>   <Source>       Scene.gd:5

for this:

```
extends Node2D

func _init():
    log_debug("MyCategory", "doing some stuff involving MyCategory")
```

I can see logging output in Creator's output panel, but not in Godot's.

File: modules/gdscript/gdscript_parser.cpp [hunk 1]
```diff
@@ -72,6 +72,53 @@ bool GDScriptParser::_end_statement() {
...
+		tokenizer->advance();
+
+		Vector<Node *> args;
+		const bool result = _parse_arguments(p_block, args, p_static);
+		if (!result) {
+			return;
+		}
+
+		if (args.size() != 2) {
+			_set_error("debug_log expects two arguments: category and message");
```
Comment (Calinou) [inline modules/gdscript/gdscript_parser.cpp]: Remember to update the messages to remove references to `debug_log` and instead refer to the correct logging function :slightly_smiling_face:

**Pull request closed without merging**